### PR TITLE
GO-1501 Add limit to change size

### PR DESCRIPTION
**Pull request by KirillSto**
---

- [x] I understand that contributing to this repository will require me to agree with the [CLA](https://github.com/anyproto/.github/blob/main/docs/CLA.md)

---

### What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI

### Related Tickets & Documents

https://linear.app/anytype/issue/GO-1501/limit-change-size

### Added tests?

- [ ] 👍 yes
- [x] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 [tech-docs](https://github.com/anyproto/tech-docs)
- [x] 🙅 no documentation needed

### CODE REVIEW

File: core/block/editor/basic/details.go [hunk 1]
```diff
@@ -42,6 +49,15 @@ func (bs *basic) SetDetails(ctx *session.Context, details []*pb.RpcObjectSetDeta
 	return nil
 }
 
+func checkDetailSizes(details []*pb.RpcObjectSetDetailsDetail) error {
+	for _, d := range details {
+		if d.Size() > detailSizeLimit {
```
[the remark this answers is not in the document]
Comment (KirillSto) [reply core/block/editor/basic/details.go]: I have rewritten this check, so we shorten the details, but not send the error

File: core/block/editor/state/normalize.go [hunk 2]
```diff
@@ -24,9 +28,17 @@ type Normalizable interface {
 }
 
 func (s *State) normalize(withLayouts bool) (err error) {
-	if err = s.Iterate(func(b simple.Block) (isContinue bool) {
+	if iErr := s.Iterate(func(b simple.Block) (isContinue bool) {
+		if b.Model().Size() > blockSizeLimit {
+			err = fmt.Errorf("size of block '%s' (%d) is above the limit of %d", b.Model().Id, b.Model().Size(), blockSizeLimit)
```
Comment (deff7) [inline core/block/editor/state/normalize.go:33]: memoize Size() call

File: core/block/editor/state/normalize.go [hunk 3]
```diff
@@ -292,6 +304,48 @@ func (s *State) normalizeSmartBlock(b simple.Block) {
 	}
 }
 
+func (s *State) shortenDetailsToLimit(details map[string]*types.Value) {
+	for key, value := range details {
+		if value.Size() > detailSizeLimit {
```
Comment (fat-fellow) [inline core/block/editor/state/normalize.go:309]: idem

File: core/block/source/source.go [hunk 4]
```diff
@@ -27,10 +27,13 @@ import (
 	"github.com/anyproto/anytype-heart/util/slice"
 )
 
+const changeSizeLimit = 10 * 1024 * 1024
```
Comment (fat-fellow) [inline core/block/source/source.go:30]: 10 mb? isn't it too loose constraint?

File: core/block/source/source.go [hunk 5]
```diff
@@ -249,6 +252,11 @@ func (s *source) PushChange(params PushChangeParams) (id string, err error) {
 	if err != nil {
 		return
 	}
```
Comment (fat-fellow) [inline core/block/source/source.go:254]: can we test it?

File: core/block/editor/state/normalize.go [hunk 6]
```diff
@@ -24,9 +28,17 @@ type Normalizable interface {
 }
 
 func (s *State) normalize(withLayouts bool) (err error) {
-	if err = s.Iterate(func(b simple.Block) (isContinue bool) {
```
Comment (fat-fellow) [inline core/block/editor/state/normalize.go:27]: I would find useful to extract a few separate functions from this method:
something like 
normalizeSize
removeInvalidChildren
normalizeRows
normalizeLayouts

and made a unittest at least for this normalizeSize case

File: core/block/source/source.go (hunk 4 above)
Comment (KirillSto) [reply core/block/source/source.go:30]: This question is open for discussion. Value of 10 Mb is stated in the ticket, and moreover @cheggaaa said we can pass up to 256 Mb as a change. Changes with higher sizes will be facing issues while networking

File: core/block/editor/state/normalize.go (hunk 2 above)
Comment (KirillSto) [reply core/block/editor/state/normalize.go]: Done!

File: core/block/editor/state/normalize.go (hunk 3 above)
Comment (KirillSto) [reply core/block/editor/state/normalize.go]: Done!

File: core/block/source/source.go (hunk 5 above)
Comment (KirillSto) [reply core/block/source/source.go:254]: Yeah, imported any-sync mocks, so could be tested easily!

File: core/block/editor/state/normalize.go (hunk 6 above)
Comment (KirillSto) [reply core/block/editor/state/normalize.go:27]: Made normalize shorted and added test on normalizeSize

File: core/block/source/source_test.go [hunk 7]
```diff
@@ -42,3 +56,108 @@ func Test_snapshotChance2(t *testing.T) {
 	// here is an example of distribution histogram
 	// https://docs.google.com/spreadsheets/d/1xgH7fUxno5Rm-0VEaSD4LsTHeGeUXQFmHsOm29M6paI
 }
+
+func TestSource_PushChange(t *testing.T) {
```
Comment (deff7) [inline core/block/source/source_test.go:60]: If you extract checking the size of snapshot into a separate function, you don't need such complicated testing code

File: core/block/source/source_test.go [hunk 8]
```diff
@@ -42,3 +56,108 @@ func Test_snapshotChance2(t *testing.T) {
...
+		return
+	}
+
+	t.Run("big change", func(t *testing.T) {
+		//given
+		_, s := initTreeAndSource()
+		params := PushChangeParams{
+			State: state.NewDoc("", nil).NewState(),
+			Changes: []*pb.ChangeContent{{&pb.ChangeContentValueOfRelationAdd{RelationAdd: &pb.ChangeRelationAdd{
+				RelationLinks: []*model.RelationLink{{Key: strings.Repeat("r", changeSizeLimit)}}},
```
Comment (deff7) [inline core/block/source/source_test.go:78]: So heavy tests. Extract size checking into separate function with maxSize parameter and you don't need to allocate so much memory

**Review by deff7 — CHANGES_REQUESTED** [review]

File: core/block/source/source.go (hunk 5 above)
Comment (deff7) [reply core/block/source/source.go:254]: I don't think importing mocks is good idea. You can use mockery to easily generate these mocks and thereby reducing dependencies. Even better you can restructure code so it have transparent business logic without dependencies at all

File: core/block/source/source_test.go (hunk 7 above)
Comment (KirillSto) [reply core/block/source/source_test.go]: Thank you for advice! It simplifies drastically

File: core/block/source/source_test.go (hunk 8 above)
Comment (KirillSto) [reply core/block/source/source_test.go]: Done!

**Review by fat-fellow — APPROVED** [review]

File: core/block/editor/state/normalize.go [hunk 9]
```diff
@@ -292,6 +321,49 @@ func (s *State) normalizeSmartBlock(b simple.Block) {
...
+		size := value.Size()
+		if size > detailSizeLimit {
+			log.With("objectID", s.rootId).Errorf("size of '%s' detail (%d) is above the limit of %d. Shortening it",
+				key, size, detailSizeLimit)
+			value, _ = shortenValueOnN(value, size-detailSizeLimit)
+		}
+	}
+}
+
+func shortenValueOnN(value *types.Value, n int) (result *types.Value, left int) {
```
Comment (deff7) [inline core/block/editor/state/normalize.go:335]: Does this function try to remove N bytes from value?

Comment (KirillSto) [reply core/block/editor/state/normalize.go:335]: Yes

File: core/block/editor/state/change.go [hunk 10]
```diff
@@ -81,6 +81,10 @@ func NewDocFromSnapshot(rootId string, snapshot *pb.ChangeSnapshot, opts ...Snap
 		pbtypes.NormalizeStruct(detailsToSave)
 	}
 
+	if detailsToSave != nil && detailsToSave.Fields != nil {
```
[the remark this answers is not in the document]
Comment (KirillSto) [reply core/block/editor/state/change.go:84]: Yeah, made it a package-level private function for now

File: core/block/editor/state/normalize.go [hunk 11]
```diff
@@ -292,6 +321,49 @@ func (s *State) normalizeSmartBlock(b simple.Block) {
...
+			newValue, n = shortenValueOnN(v, n)
+			value.GetListValue().Values[i] = newValue
+			if n == 0 {
+				return value, 0
+			}
+		}
+		return value, n
+	case *types.Value_StructValue:
+		var newValue *types.Value
+		for key, v := range value.GetStructValue().GetFields() {
```
[the remark this answers is not in the document]
Comment (KirillSto) [reply core/block/editor/state/normalize.go]: I have tried to introduce sorting of struct fields and values of lists. Benchmark showed that shortenValueOnN is working 2 times slower and consumes 50% more heap, Howevwr, if we assume that invokation of this func is a rare event, moreover with big amount of fields/values, than it is ok imho

File: core/block/editor/state/normalize.go [hunk 12]
```diff
@@ -292,6 +321,49 @@ func (s *State) normalizeSmartBlock(b simple.Block) {
...
+		if size > detailSizeLimit {
+			log.With("objectID", s.rootId).Errorf("size of '%s' detail (%d) is above the limit of %d. Shortening it",
+				key, size, detailSizeLimit)
+			value, _ = shortenValueOnN(value, size-detailSizeLimit)
+		}
+	}
+}
+
+func shortenValueOnN(value *types.Value, n int) (result *types.Value, left int) {
+	switch value.Kind.(type) {
```
[the remark this answers is not in the document]
Comment (KirillSto) [reply core/block/editor/state/normalize.go]: Fixed

File: core/block/editor/state/change.go [hunk 13]
```diff
@@ -251,6 +251,7 @@ func (s *State) changeBlockDetailsSet(set *pb.ChangeDetailsSet) error {
 			Fields: make(map[string]*types.Value),
 		}
 	}
+	shortenDetailsToLimit(s.rootId, map[string]*types.Value{set.Key: set.Value})
```
Comment (deff7) [inline core/block/editor/state/change.go:254]: It's a little counterintuitive that Value is being shorted after calling `shortenDetailsToLimit`. Maybe extract the part of this function that handles the single value?

File: core/block/editor/state/normalize.go (hunk 9 above)
Comment (deff7) [reply core/block/editor/state/normalize.go:337]: shortenOn sounds like we short value **on** specific position. Maybe use **by** preposition?

File: core/block/editor/state/change.go (hunk 13 above)
Comment (KirillSto) [reply core/block/editor/state/change.go]: Yeah, as we assume that each value should be limited to 64 kilobytes, pull out single-value logic to separate function

**Review by fat-fellow — APPROVED** [review]

**Review by deff7 — APPROVED** [review]

**Pull request merged**